### PR TITLE
[tapocontrol] new communication protocol integration / code revision

### PR DESCRIPTION
**Pull request by wildcs**
**Tapocontrol-Binding for Tapo-Devices**

this pr is necesarry because manufacturer changed communication protocol for some devices so device-communication is broken.
because the communication changed i decided to revise the whole code of the plugin.

this pr has more changes:

- complete code revision
- integration new "klap-protocol"
- integration local device discovery with udp
- integration "H100 Smart-Hub"
- integration "T110 door-/window-contact sensor"
- integration "T310/T315 wheateher- sensors"
- supporting light-effects for light-strips

### CODE REVIEW

File: bundles/org.openhab.binding.tapocontrol/pom.xml [hunk 1]
```diff
@@ -14,4 +14,12 @@
 
   <name>openHAB Add-ons :: Bundles :: TapoControl Binding</name>
 
+  <dependencies>
+    <!-- keep until core pom is equal or newer - for support of gson-records as needed -->
+    <dependency>
+      <groupId>com.google.code.gson</groupId>
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/pom.xml:20]: This is no longer needed. gson 2.10 is now part of the distro.

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/TapoProtocolEnum.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            BY_VALUE.put(e.value, e);
+        }
+    }
+
+    private TapoProtocolEnum(String value) {
+        this.value = value;
+    }
+
+    public static TapoProtocolEnum valueOfString(String label) {
+        return Objects.requireNonNull(BY_VALUE.get(label));
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/TapoProtocolEnum.java:45]: You could avoid the usage of the additional Map, with something like : 
`EnumSet.allof(TapoProtcolEnum.class).stream().filter(p -> p.value.equals(label)).findfirst().orElseThrow(() -> new IllegalArgumentException());`

(uncontrolled code - may contain errors)

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/TapoProtocolInterface.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ * @author Christian Wild - Initial contribution
+ */
+@NonNullByDefault
+public interface TapoProtocolInterface {
+
+    /***********************
+     * Init Class
+     **********************/
+
+    /*
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/TapoProtocolInterface.java:34]: Should be removed

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/klap/KlapCipher.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+ * @author Christian Wild - Initial contribution
+ */
+@NonNullByDefault
+public class KlapCipher {
+    private int ivSeq = 0;
+    private byte[] iv;
+    private byte[] key;
+    private byte[] sig;
+
+    /* class contants */
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/klap/KlapCipher.java:41]: Maybe just cosmetic but I like to see static final at the very beginning of the class.

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/klap/KlapProtocol.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+            throw new TapoErrorHandler(ERR_BINDING_HTTP_RESPONSE, "invalid response receicved");
+        }
+    }
+
+    /**
+     * Get Tapo-Response from responsestring
+     * decrypt if is encrypted
+     */
+    protected TapoResponse getTapoResponse(String responseString) throws TapoErrorHandler {
+
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/klap/KlapProtocol.java:238]: Remove empty line. Have you "mvn spotless:apply" ?

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/passthrough/PassthroughProtocol.java [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+import com.google.gson.GsonBuilder;
+
+/**
+ * Handler class for TAPO-PASSTHROUGH-Protocol
+ *
+ * @author Christian Wild - Initial contribution
+ */
+@NonNullByDefault
+public class PassthroughProtocol implements TapoProtocolInterface {
+    protected static final Gson gson = new GsonBuilder().disableHtmlEscaping().excludeFieldsWithoutExposeAnnotation()
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/passthrough/PassthroughProtocol.java:52]: Can't 'gson' be defined at `TapoControlHandlerFactory` once for all the binding ?

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/passthrough/PassthroughProtocol.java [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+            .create();
+
+    private final Logger logger = LoggerFactory.getLogger(PassthroughProtocol.class);
+    private final TapoConnectorInterface httpDelegator;;
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/passthrough/PassthroughProtocol.java:56]: Superfluous `;`

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/securePassthrough/SecurePassthrough.java [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+import com.google.gson.GsonBuilder;
+
+/**
+ * Handler class for TAPO-SECUREPASSTHROUGH-Protocol
+ *
+ * @author Christian Wild - Initial contribution
+ */
+@NonNullByDefault
+public class SecurePassthrough implements TapoProtocolInterface {
+    private static final Logger logger = LoggerFactory.getLogger(SecurePassthrough.class);
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/protocol/securePassthrough/SecurePassthrough.java:51]: `logger` isn't static.

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/bridge/TapoBridgeHandler.java [hunk 9]
```diff
@@ -114,7 +114,7 @@ public void dispose() {
      */
     @Override
     public Collection<Class<? extends ThingHandlerService>> getServices() {
-        return Set.of(TapoDiscoveryService.class);
+        return Collections.singleton(TapoDiscoveryService.class);
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/bridge/TapoBridgeHandler.java:117]: `Set.of` was fine

**Review by clinique — CHANGES_REQUESTED** [review]
Nice rework. Thanks for this !

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/TapoControlHandlerFactory.java [hunk 10]
```diff
@@ -42,15 +46,20 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
 /**
- * The {@link ThingHandler} is responsible for handling commands, which are
+ * The {@link TapoControlHandler} is responsible for handling commands, which are
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/TapoControlHandlerFactory.java:53]: Wrong comment header, you should be presenting `TapoControlHandlerFactory` here.

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/TapoCloudConnector.java [hunk 11]
```diff
@@ -13,226 +13,179 @@
...
-        this.token = getToken(username, password, UUID.randomUUID().toString());
-        this.url = TAPO_CLOUD_URL + "?token=" + token;
-        return !this.token.isBlank();
+    public void handleResponse(TapoResponse tapoResponse, String command) throws TapoErrorHandler {
+        switch (command) {
+            case CLOUD_CMD_LOGIN:
+                handleLoginResult(tapoResponse);
+                break;
+            case CLOUD_CMD_GETDEVICES:
+                bridge.getDiscoveryService().addScanResults((getDeviceListFromResponse(tapoResponse)));
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/api/TapoCloudConnector.java:79]: `.addScanResults((getDeviceListFromResponse(tapoResponse)));` => `.addScanResults(getDeviceListFromResponse(tapoResponse));`

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/TapoControlHandlerFactory.java [hunk 12]
```diff
@@ -42,15 +46,20 @@
...
  * sent to one of the channels.
  *
  * @author Christian Wild - Initial contribution
  */
 @Component(service = ThingHandlerFactory.class, configurationPid = "binding.tapocontrol")
 @NonNullByDefault
 public class TapoControlHandlerFactory extends BaseThingHandlerFactory {
+    public static final Gson GSON = new GsonBuilder().disableHtmlEscaping().excludeFieldsWithoutExposeAnnotation()
+            .create();
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/TapoControlHandlerFactory.java:62]: Using `.setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)` you can save a lot of annotations.

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/dto/TapoLightEffect.java [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+@NonNullByDefault
+public class TapoLightEffect {
+    @Expose
+    private String id = "";
+
+    @Expose
+    private String name = "";
+
+    @Expose
+    private Integer brightness = 0;
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/dto/TapoLightEffect.java:46]: `int` instead of `Integer` ?

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/dto/TapoLightEffect.java [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+
+    @Expose
+    @Nullable
+    @SerializedName("display_colors")
+    private List<int[]> displayColors = List.of();
+
+    @Expose
+    private Integer enable = 0;
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/dto/TapoLightEffect.java:54]: See above

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/dto/TapoLightEffect.java [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+
+    @Expose
+    @Nullable
+    private Integer bAdjusted = 0;
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/dto/TapoLightEffect.java:58]: Same here

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/rf/TapoChildDeviceHandler.java [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+     * INITIALIZE DEVICE
+     */
+    @Override
+    public void initialize() {
+        logger.trace("({}) Initializing thing ", uid);
+        try {
+            Bridge bridgeThing = getBridge();
+            if (bridgeThing != null) {
+                BridgeHandler bridgeHandler = bridgeThing.getHandler();
+                if (bridgeHandler != null) {
```
Comment (clinique) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/rf/TapoChildDeviceHandler.java:82]: You could write `if (bridgeHandler instanceof TapoHubHandler tapoHubHandler)`

**Review by clinique — CHANGES_REQUESTED** [review]
A small review pass.

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/TapoControlHandlerFactory.java (hunk 12 above)
Comment (wildcs) [reply bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/TapoControlHandlerFactory.java:62]: Maybe i save some annotiations with underlines but i had to add others witch camelCase because it tapo used mixed FieldNamePolicies.

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/dto/TapoLightEffect.java (hunk 13 above)
Comment (wildcs) [reply bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/dto/TapoLightEffect.java:46]: Integer is used because @Nullable is necesarry to send null in request for function of the "set request".

**Review by clinique — APPROVED** [review]
LGTM

**Comment by lsiepel** [comment]
@wildcs can you fix the conflict?

**Comment by wildcs** [comment]
> @wildcs can you fix the conflict?

Can youn tell me how? The code has changed completeley. The old file has to be overwritten by the new one file.  I dont have write access to do this

**Comment by clinique** [comment]
> Can someone else help to resolve the conflict?
> @clinique @lolodomo  maybe?

You need to rebase your PR and force push, it should do it.

**Comment by mccaffm** [comment]
> Hello everybody. Please take a look on this request. Many people waiting for this to be merged. ;-) Thanks by advance!

does anyone know who/how to prompt openhab/add-ons-maintainers all the documentation re maintainers sees to be removed in 4.1.0 and the 4.0.1 links return 404 for me.? @clinique - gael i assume you are a maintainer- but can you merge after having done the review, its not clear anymore to a lowly user.

**Comment by clinique** [comment]
> > Hello everybody. Please take a look on this request. Many people waiting for this to be merged. ;-) Thanks by advance!
> 
> does anyone know who/how to prompt openhab/add-ons-maintainers all the documentation re maintainers sees to be removed in 4.1.0 and the 4.0.1 links return 404 for me.? @clinique - gael i assume you are a maintainer- but can you merge after having done the review, its not clear anymore to a lowly user.

I'm not a maintainer, only doing some code reviews from time to time. I know it works and will try to ping some maintainers so they can have a look ahead of the 4.1 final release. @lolodomo , @J-N-K , @jlaur  if someone of you can have a look at it ?

**Comment by mccaffm** [comment]
> > > Hello everybody. Please take a look on this request. Many people waiting for this to be merged. ;-) Thanks by advance!
> > 
> > 
> > does anyone know who/how to prompt openhab/add-ons-maintainers all the documentation re maintainers sees to be removed in 4.1.0 and the 4.0.1 links return 404 for me.? @clinique - gael i assume you are a maintainer- but can you merge after having done the review, its not clear anymore to a lowly user.
> 
> I'm not a maintainer, only doing some code reviews from time to time. I know it works and will try to ping some maintainers so they can have a look ahead of the 4.1 final release. @lolodomo , @J-N-K , @jlaur if someone of you can have a look at it ?

thankyou @clinique like i said, its difficult to understand current mmebers and process. <M>

**Review by holgerfriedrich — CHANGES_REQUESTED** [review]
@wildcs thanks for the contribution! I don't not anything about tapocontrol and I am no maintainer as well..... 

But I can give a observations on code style etc.:
* please resolve javadoc errors: ```mvn javadoc:javadoc``` should not show errors (warnings about missing docs are fine)
* consider renaming folder (and the package name in all files below) ```internal/api/protocol/securePassthrough``` to lowercase - otherwise it will cause SAT warnings
* do a ```mvn verify``` and check the SAT report in the target folder
* see comments below
* It seems there are **no tests at all** (or did I overlook them)
Remark: I did not carefully review this PR and cannot judge technical details

**Comment by wildcs** [comment]
@holgerfriedrich 
Thanks for your review. Is solved your change requests.
mvn javadoc:javadoc has now only warnings. mvn clean install is working fine for me. I don't know why Jenkins fails here now.

**Comment by holgerfriedrich** [comment]
Please try to rebase. We have updated base libraries, thus the build is failing.

**Comment by wildcs** [comment]
@holgerfriedrich 
ok thanks 👍 . now it should be done so far

File: bundles/org.openhab.binding.tapocontrol/pom.xml [hunk 17]
```diff
@@ -13,5 +13,4 @@
   <artifactId>org.openhab.binding.tapocontrol</artifactId>
 
   <name>openHAB Add-ons :: Bundles :: TapoControl Binding</name>
-
```
Comment (holgerfriedrich) [inline bundles/org.openhab.binding.tapocontrol/pom.xml:16]: you may undo this change

**Review by holgerfriedrich — COMMENTED** [review]
Remark: I did not carefully review this PR and cannot judge technical details

Thanks, my main points (SAT+Javadoc) have been addressed.

Except this one:
It seems there are no tests at all (or I have overlooked them) - out of scope of this PR?

Could you explain why there is not test at all? The existing tests have been deleted.
Or did I get something wrong?

File: bundles/org.openhab.binding.tapocontrol/pom.xml (hunk 17 above)
Comment (wildcs) [reply bundles/org.openhab.binding.tapocontrol/pom.xml]: It's impossible to check and judge all technical deteails and functions in a large repo like oh. 
So you can just look if code is clean and looks fine. So thanks you for doing this.

What tests do you mean? 
If you mean the /test/ folder this was still a storage for code i tried some functions (like different device-discovery methods).
So it was no usefull stuff there at any time.

Comment (holgerfriedrich) [reply bundles/org.openhab.binding.tapocontrol/pom.xml]: I was referring to unit tests which are part of the CI run (mvn verify).
https://www.openhab.org/docs/developer/tests.html

**Review by holgerfriedrich — APPROVED** [review]
Just checked warnings, SAT, javadoc, i18n -> 👍 from my side

**Comment by wildcs** [comment]
Hey Guys, had to do rebase and resolve confilcts manually because of core changes again.
I know you have a lot of work and many PR to review and it hink you do this in your free time too. Thanks therefore.
This PR contains a lot of changes to the origin but it was necesarry because of a protocol change by the device-vendor. So the old binding isnt working anymore for the most devices.
I'm happy to continue to develop the binding, but it's always more work to keep this pr up to date on core changes.

**Comment by dor-bw** [comment]
Hello all. I hope this fix is merged very soon. I'm waiting for so long to get my items work again in my openhab installation.
@lolodomo @jlaur please look at these MR!!!
Thanks by advance!

**Comment by gelsas** [comment]
When will this be pushed? It seems everything is fine now, right?

**Comment by clinique** [comment]
@lolodomo : sorry to ping you, can you help having this merged ?

File: bundles/org.openhab.binding.tapocontrol/README.md [hunk 18]
```diff
@@ -43,21 +56,32 @@ This is used for device discovery and to create a handshake (cookie) to act with
...
+|------------------------|-----------------------------------------------------------------------------------------------------|
+| username               | Username (eMail) of your Tapo-Cloud                                                                 |
+| password               | Password of your Tapo-Cloud                                                                         |
+| cloudDiscovery         | Use Cloud Discovery-Service to get all in Tapo-App registered devices. Includes DeviceName. IP-Address and Encryption has to set manually                     |
+| udpDiscovery           | Use UDP Discovery-Service to discover online devices in the local network. Includes Encryption and IP-Address. Results will be merged with cloud discovery    |
+| onlyLocalOnlineDevices | [advanced] Uses Cloud and UPD-Discovery to get more informations but will only discover online devices via UDP |
+| broadcastAddress       | [advanced] Set broadcast address to your local subnet  if you have problems with default address   |
+| discoveryInterval      | [advanced] Interval in minutes when a background device scan should be executed. Default is 60     |
```
Comment (lsiepel) [inline bundles/org.openhab.binding.tapocontrol/README.md:67]: Minor: can you fix the table markdown?

File: bundles/org.openhab.binding.tapocontrol/README.md [hunk 19]
```diff
@@ -43,21 +56,32 @@ This is used for device discovery and to create a handshake (cookie) to act with
...
 
-## Thing Configuration
 
-The thing needs to be configured with `ipAddress`.
+## Thing Configurations
```
Comment (lsiepel) [inline bundles/org.openhab.binding.tapocontrol/README.md:70]:
```suggestion
## Thing Configuration
```

File: bundles/org.openhab.binding.tapocontrol/README.md [hunk 20]
```diff
@@ -72,12 +96,20 @@ All devices support some of the following channels:
...
 |           | colorTemperature | Number                 | White-Color-Temp 2500-6500K         | L510, L530, L610, L630, L900, L920                               |
 |           | color            | Color                  | Color                               | L530, L630, L900, L920                                           |
-| effects   | fxName           | String                 | Active lightning effect             | L530                                                             |
+| sensor    | isOpen           | Switch                 | Contact (Door/Window) is Open       | T110                                                             |
+|           | currentTemp      | Number:Temperature     | Current Temperature                 | T310, T315                                                       |
+|           | currentHumidity  | Number:Dimensionless   | Current relative humidity in %      | T310, T315                                                       |
+| effects   | fxName           | String                 | Active lightning effect             | L530, L900, L920, L930                                                           |
```
Comment (lsiepel) [inline bundles/org.openhab.binding.tapocontrol/README.md:102]: Minor: can you fix the table markdown?

File: bundles/org.openhab.binding.tapocontrol/README.md [hunk 21]
```diff
@@ -43,21 +56,32 @@ This is used for device discovery and to create a handshake (cookie) to act with
...
+| broadcastAddress       | [advanced] Set broadcast address to your local subnet  if you have problems with default address   |
+| discoveryInterval      | [advanced] Interval in minutes when a background device scan should be executed. Default is 60     |
 
-## Thing Configuration
 
-The thing needs to be configured with `ipAddress`.
+## Thing Configurations
 
-The thing has the following configuration parameters:
+WiFi-Things needs to be configured with `ipAddress`.
```
Comment (lsiepel) [inline bundles/org.openhab.binding.tapocontrol/README.md:72]:
```suggestion
WiFi-based things needs to be configured with `ipAddress`.
```

File: bundles/org.openhab.binding.tapocontrol/README.md [hunk 22]
```diff
@@ -43,21 +56,32 @@ This is used for device discovery and to create a handshake (cookie) to act with
...
+RF-Things need a Hub (WiFi-Device) to operate.
```
Comment (lsiepel) [inline bundles/org.openhab.binding.tapocontrol/README.md:73]:
```suggestion
RF-based things need a SmartHub(WiFi-Device) to operate. 
```

Not sure if it should be Hub or SmartHub, or Smarthub, but it this moment this description and the table below don;t have a consistent naming for the hub

**Review by lsiepel — CHANGES_REQUESTED** [review]
Seen about  ~70 files. Most looks good, do have some comments regarding the documentation. As this is a large PR that touches many files, testing is very much needed. The startpost of this PR, does not have a jar file or link to a community topic with users testing these changes. Can you confirm this is well tested?

**Comment by dor-bw** [comment]
> @dor-bw until this PR is merged you can manually install my compiled .jar from my [oh-compiled-addons](https://github.com/wildcs/oh3_compiled-addons/tree/tapocontrol) repo.

@wildcs Hello. This don't work for me. I follow your instructions from your github site but no Binding are shown inmy openhab installation.

File: bundles/org.openhab.binding.tapocontrol/README.md (hunk 18 above)
Comment (wildcs) [reply bundles/org.openhab.binding.tapocontrol/README.md]: Tried to fix it, but because of long text it's not perfect. if you have a better idea please post it :)

**Comment by wildcs** [comment]
@lsiepel 
Thanks for reviewing. Fixed your review requests except one minor.
Yes this pr is tested well since the pr was created by many users and as you can see in some comments it's eagerly awaited 
Forgot to link the [support-thread](https://community.openhab.org/t/tapocontrol-control-tapo-smart-wifi-devices-with-openhab-official-support-thread/118389/397) in the startpost. Most problems are because of some people didn't know how to install a binding manually the right way.
The thing is: Without this pr, the binding isn't working for most devices anymore! So even if there should be some bugs it would be still better than it is now.

**Comment by clinique** [comment]
> @lsiepel 
> Thanks for reviewing. Fixed your review requests except one minor.
> Yes this pr is tested well since the pr was created by many users and as you can see in some comments it's eagerly awaited 
> Forgot to link the [support-thread](https://community.openhab.org/t/tapocontrol-control-tapo-smart-wifi-devices-with-openhab-official-support-thread/118389/397) in the startpost. Most problems are because of some people didn't know how to install a binding manually the right way.
> The thing is: Without this pr, the binding isn't working for most devices anymore! So even if there should be some bugs it would be still better than it is now. 

I second this.

**Comment by nelsonaponte** [comment]
@lsiepel, I have been using the updated binding from the support thread since the 4th of December without problems to command my devices with the latest firmware (using the new KLAP protocol).
This PR also fixes a problem I reported on a separate [issue](https://github.com/openhab/openhab-addons/issues/15717), as you can see from the comments.

**Comment by hmerk** [comment]
@nelsonaponte 
Sorry Nelson, but if it is that urgent, contributers/developers should publish test versions of bindings through our marketplace. This is one of the reasons it was created for.
It does not make sense if all users comment here.

**Comment by nelsonaponte** [comment]
Hi, @hmerk.
That goes both ways. If the marketplace is so important for that, I hope it's specified somewhere (_easy to find_) so developers know about it. I'm not one but I couldn't find any mention of the "Marketplace" in the Documentation (Developer Guide and Developer Appendix) or even the Add-On Marketplace itself.

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/wifi/TapoBaseDeviceHandler.java [hunk 23]
```diff
@@ -85,23 +86,27 @@ protected TapoDevice(Thing thing) {
...
+
+    /**
+     * Init TapoBridgeHandler TapoDeviceConnector
+     */
+    protected void initConnector() {
+        Bridge bridgeThing = getBridge();
+        if (bridgeThing != null) {
+            BridgeHandler bridgeHandler = bridgeThing.getHandler();
+            if (bridgeHandler != null) {
+                this.bridge = (TapoBridgeHandler) bridgeHandler;
```
Comment (lsiepel) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/wifi/TapoBaseDeviceHandler.java:107]:
```suggestion
            if (bridgeHandler instanceof TapoBridgeHandler tapoBridgeHandler) {
                this.bridge = tapoBridgeHandler;
```

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/wifi/TapoBaseDeviceHandler.java [hunk 24]
```diff
@@ -132,29 +137,39 @@ private void activateDevice() {
...
+            return false;
         }
-        return configErr;
+        return true;
+    }
+
+    /**
+     * Check Device Configuration
+     */
+    protected boolean checkConfig() {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/wifi/TapoBaseDeviceHandler.java:159]: I personally prefer to make the configuration class responsible to check if it is valid or not. 
Maybe something to think about, it would make this class also more readable.

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/wifi/bulb/TapoBulbHandler.java [hunk 25]
```diff
@@ -0,0 +1,262 @@
...
+     * Set light effect
+     * 
+     * @param fxId (String) id of LightEffect
+     */
+    protected void setLightEffect(String fxId) {
+        try {
+            TapoLightDynamicFx fxData = new TapoLightDynamicFx();
+            fxData.setEffect(fxId);
+            connector.sendCommandAndQuery(DEVICE_CMD_SET_DYNAIMCLIGHT_FX, fxData, bulbData.supportsMultiRequest());
+        } catch (Exception e) {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/wifi/bulb/TapoBulbHandler.java:211]: Would it be possible to use checked exceptions?

**Review by lsiepel — CHANGES_REQUESTED** [review]
Looked at some more files. Sorry lack of time to complete my review, but don't want to hold back these comments.

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/wifi/TapoBaseDeviceHandler.java (hunk 24 above)
Comment (wildcs) [reply bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/devices/wifi/TapoBaseDeviceHandler.java:159]: Good idea. I will change it

**Comment by wildcs** [comment]
> @nelsonaponte Sorry Nelson, but if it is that urgent, contributers/developers should publish test versions of bindings through our marketplace. This is one of the reasons it was created for. It does not make sense if all users comment here.

This wasn't clear for me too. I published it in the support-thread of the binding oh oh-community

**Comment by hmerk** [comment]
So you all have missed this announcement and many published bindings on the marketplace 😉

https://community.openhab.org/t/about-the-add-on-marketplace-category/123408

**Comment by moodyblue** [comment]
Pls put some urgency on this request.

File: bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/helpers/utils/TapoUtils.java [hunk 26]
```diff
@@ -0,0 +1,187 @@
...
+            } else if (SUPPORTED_MOTION_SENSORS.contains(deviceUID)) {
+                deviceLabel = DEVICE_DESCRIPTION_MOTION_SENSOR;
+            } else if (SUPPORTED_WHEATHER_SENSORS.contains(deviceUID)) {
+                deviceLabel = DEVICE_DESCRIPTION_TEMP_SENSOR;
+            }
+            if (alias.length() > 0) {
+                return String.format("%s %s %s (%s)", DEVICE_VENDOR, deviceModel, deviceLabel, alias);
+            }
+            return String.format("%s %s %s", DEVICE_VENDOR, deviceModel, deviceLabel);
+
```
Comment (lsiepel) [inline bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/helpers/utils/TapoUtils.java:182]:
```suggestion
```

**Review by lsiepel — CHANGES_REQUESTED** [review]
Very minor comment, otherwise LGTM

**Comment by wildcs** [comment]
@lsiepel 
Thanks for your review. Did a rebase and fixed your request.
Hope it now can be merged soon.

**Comment by clinique** [comment]
@holgerfriedrich : can you please push on having it merged soon ? Thanks a lot.

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM

**Comment by dor-bw** [comment]
Thank you very much! All good things are worth waiting for.

**Comment by nelsonaponte** [comment]
> Thank you very much! All good things are worth waiting for.

That’s one funny way to end this drama 😅

@wildcs, thanks for not giving up!
@clinique and @lsiepel, thanks for your support!

**Comment by giig1967g** [comment]
In which version will we see the new Tapo binding?

**Comment by lsiepel** [comment]
> In which version will we see the new Tapo binding?

Daily snapshot, next milestone 4.2.0.M3 or next stable 4.2.0
